### DISCUSSION TITLE
Sops SecretGenerator to create secrets from sops-encrypted files during resource generation

### DISCUSSION THREAD

**Discussion by scp7** [fluxcd/flux2#745] [Q&A] [answered]
Guys - great job with flux2, been using in anger now for a week or so and very positive experience. 

I have a terraform workflow for generating GKE infrastructure and am templating helm values files for use in flux as part of that. What I want to do is generate configmaps and secrets from these values files and reference them as part of the helm install. 

This works successfully, for example I can consume values from files using the generator methods that have been explained in the tutorial, e.g. here's an example kustomisation I'm using.

```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: postgresql
resources:
  - ../../../base/postgresql
configMapGenerator:
  - name: values-postgresql
    files:
      - values.yaml=values-postgresql.yaml
secretGenerator:
  - name: secrets-postgresql
    files:
      - values.yaml=secrets-postgresql.yaml
```

Now ideally want I want to do is decrypt the secrets-postgresql.yaml (typical helm values file) file as part of the generator action. From trial and error I believe the current decryption only works on a statically encrypted kubernetes secret (where just the data section is encrypted whereas I'd need the values file to be descrypted as part of the generation.

I think I'm looking for flux2 support for something like this: https://github.com/goabout/kustomize-sopssecretgenerator 

Any ideas?

**Comment by stefanprodan** [comment]
Flux2 has native support for SOPS, here are the docs: https://toolkit.fluxcd.io/guides/mozilla-sops/

You would encrypt the secrets with SOPS CLI, include them in your` kustomization.yaml` as resources and place them in your Git repo. Then use the secrets in the HelmRelease like this: https://toolkit.fluxcd.io/guides/helmreleases/#refer-to-values-in-configmap-and-secret-resources

**Reply by bob-rohan** [reply]
Hey @stefanprodan , your suggestion negates the use of `secretGenerator`, I think? 

secretGenerator is useful mechanism to automajically rollout changes across pods referencing changed secrets, employing hash of content as a name suffix to both secret resource and references to that secret. Of this all I'm sure you're aware, just adding for context and others.

I've read a few posts suggesting use of .flux.yaml `command: for file in secrets/*; do echo "sops -d -i $file"; sops -d -i "$file"; done;`

But read .flux.yaml was deprecated in v2, I think?

I have read to include this step as part of CI, sops -d .env file or suchlike, creating Secret as you've described, then sops -e , commit result. Not sure I like the idea of CI having this access. Certainly don't like the thought of having end users maintain this locally either. Would prefer SOPS to remain only where required, end-user machine, in cluster...

Do you have thoughts on how I might achieve something similar to secretsGenerator? Link provided by @scp7 suggests plugins may be available, although I'm not sure how that's used when kustomize is running from flux provided source. Guess would need to fork the flux kustomize controller. Not particularly wanting to do that, nor particularly hoping to use something outside the flux eco-system for security..  

For reference https://github.com/fluxcd/flux/issues/3433

**Reply by kingdonb** [reply]
I think that's a great example for manifest generation, if it's possible to use secretGeneration like you described. You have an example of how to do this in `.flux.yaml`, but now with SOPS, I think we use `.sops.yaml`

Can kustomize-controller decrypt files that are not Kubernetes yaml? I think it will ignore any file that does not have a `.yaml` extension at apply time, which is good for `.env` becaue you don't want anything applied to the cluster that is not valid K8s yaml (it will cause the whole Kustomization to fail.)

I think this will work, if so perhaps there is a little bit more information we can add to the guide near this point to make it easier to discover this functionality:

https://toolkit.fluxcd.io/guides/mozilla-sops/#configure-the-git-directory-for-encryption

I'm still working through the sops tutorial myself, and this is a bit above and beyond what is described, but at this point in the guide it is explained that you must write a file called `.sops.yaml` – this is a standard SOPS configuration file, as described there under the heading:

> **Using .sops.yaml conf to select KMS/PGP for new files**

This is the example provided in the Flux guide:

```
cat <<EOF > ./clusters/cluster0/.sops.yaml
creation_rules:
  - path_regex: .*.yaml
    encrypted_regex: ^(data|stringData)$
    pgp: ${KEY_FP}
EOF
```

So, if there were other files that you wanted to decrypt, I gather this would be the place to let SOPS know that it should try decrypting those other files than just "anything `.yaml`" which has an encrypted `data` or `stringData`.

The details about `creation_rules` can be found here:
https://autrilla.gitbooks.io/sops/content/uage.html

The SOPS `README.md` also carries a similar example, but this one is better – shows how it can be used both with and without `encrypted_regex:`, I presume that leaving this field off will make SOPS aware that it should decrypt the whole file.

Did you try adding another rule here for `.env` by any chance? Something like this?

I am not sure of any reason why this won't work, but I haven't had enough time to complete the guide and modify it for this yet. Maybe later this morning, if I can get around to it before you have a chance to test this and see if it works.

```
...
    - filename_regex: \.env$
      pgp: ${KEY_FP}
EOF
```

Thanks for the suggestion @bob-rohan I will definitely be including something like this in our Manifest Generation guide, if it turns out to solve your problem!

**Reply by bob-rohan** [reply]
@kingdonb appreciate the interest, not sure where you're going with all that. I believe this is the crux

https://github.com/fluxcd/kustomize-controller/blob/198b8e7dc169c4f0c6d56e3cc4c4a7564d73995d/controllers/kustomization_decryptor.go#L77

```	
if kd.kustomization.Spec.Decryption != nil && kd.kustomization.Spec.Decryption.Provider == DecryptionProviderSOPS &&
		bytes.Contains(out, []byte("sops:")) && bytes.Contains(out, []byte("mac: ENC[")) {
```

If you'll only ever decrypt entire files, then you'll never detect encrypted content that was base64 encoded prior to being added to a secret manifest as part of the kustomize secretsgenerator.

Either base64 decode the secret data and check for `bytes.Contains(out, []byte("sops:")) && bytes.Contains(out, []byte("mac: ENC[")` or rely on some annotation `metadata.annotations.flux-decrypt-my-secret-data-with-sops-please=true

**Comment by XtremeAI** [comment]
Hi guys, 

I am also very interested in having this sorted. 
It makes perfect sense to source from a **generated** secret! 
When you do try to create say a secret for a HelmRelease out of your good looking encrypted values file:

secret-values.yaml:
```
a: ENC[
b: 
 ba: ENC[
 bc: ENC[
...
```
```
  secretGenerator:
    - name: secret
      files:
        - values.yaml=secret-values.yaml  # sops encrypted
```
it will complain about ENC[... blah blah... this says that obviously SOPS decrypt was not yet executed.

Using pre-created and SOPS encryped secrets leaves us with over obscure secrets and a requirements to re-generate them when a value changes: 
```
apiVersion: v1
data:
  values.yaml: ENC[... what the hell is in there?? 
  kind: Secret
metadata:
  name: secret
```
It would be super cool if Kustomize controller would sops decrypt the values before it does kustomize build so we could use generated secrets!

How can we achieve this?

**Reply by kingdonb** [reply]
I was able to get secretsGenerator to act on my extra encrypted secret files, but I just wind up with an encrypted secret applied (not decrypted) into the cluster.

I used `--output-type=json` and `--input-type=binary` when doing `sops -e` and the result was a secret in the cluster that has stringData looks like this:

```
application.properties: "{\n\t\"data\": \"ENC[AES256_GCM,data:Z9nNKqWHjsNherPVtvCsKw==,iv:cvv29ymrJucPX+0n7rKTD85I4BHOCBJTKGmYryL+Gvg=,tag:sscOybQH7JVAlDwy060cRA==,type:str]\",\n\t\"sops\":
  {\n\t\t\"kms\": null,\n\t\t\"gcp_kms\": null,\n\t\t\"azure_kv\": null,\n\t\t\"hc_vault\":
  null,\n\t\t\"age\": null,\n\t\t\"lastmodified\": \"2021-03-29T12:53:51Z\",\n\t\t\"mac\":
  \"ENC[AES256_GCM,data:TfQ2Oqa4hB7OyMnWKqAjh6HUymoYKNcerJ+T0H2mau732uQDqKgwL4qIMC2kEg0utQ9hU7g7/T+l2bPCbdaVU20kSCPwcGNHRQDfsYIoIwL8Ajw8zHf9ID678bvV5fn24s/T4JhJB5uoTSHn6Fb9uUcYjPHVChyxHPkT9+Yfcso=,iv:r4vPSnv/J5YHyp/eTIlEJGKhG+Jb0DyxJbViW2pvfSU=,tag:8hp/mlE3BONBVq0xR3Npvw==,type:str]\",\n\t\t\"pgp\":
  [\n\t\t\t{\n\t\t\t\t\"created_at\": \"2021-03-29T12:53:50Z\",\n\t\t\t\t\"enc\":
  \"-----BEGIN PGP MESSAGE-----\\n\\nhQIMA06rSFOpPqRMAQ/8DBLOlcf2CxSVIEVE5STWQeAbNVGqg+kE/gKxbVoKxdtt\\nvXNovGI0+VEeevc8Z2Y0yMFMs3Rzq6fHx0t3+Z41BFcz56FtvvCOEomi5r2ML24X\\n4ired56T7iI9dllTfoHwE2U9anoptfeDAlQ4sRq0ix+azbDlEUjI09j44aYlK4YJ\\nxSNynbbIslNqH8S13NTrpNjjFj2e8ceNl+SpP25UkUgxjMla1yUmqLxR3fFVT9jf\\nbvXH2oCE1XlhH6jbG8zAAPZ6ZVN9eylhOOATmEHhdj4zPiOy9l3FwXmskQwJsQHX\\nBceCpkdgs6NFIeDdMsRJ9EVU90klA/6Oy3PQIcVoLd2A4CCWp1XOHU7va2NJ3YF2\\nlYKNuPFrZ1hYlpxWKVtn994QAC6gdEQEZD2x3z+Z+CY10xgPApHesNZG2/wB1spV\\nHuFAhCm9h0qErgT4ptG+nkrZ6cnK04dGmODygg3lsFV/RTMb4NUU5nQyds9GXjYz\\nseVO4ZTBvXoGvVVN3oEODuKVJ4hajHb0pRUTMJfHYKaGU5f5XAR07i7UYN0GDQ/6\\nF6K3QN6KW0UA5mclLWSu8zYvWA3XXsThPduWxu4zh8jxol5umG1LuTIVjWXbvq9w\\nSfHY78l3l6Rj2J+V6Xbs17SmCP5GppaE4HLlusD1Dw38ftoE5bsO3/e1f83IUxbS\\nXgE9zK7A6AtT8V5IrEGagVABRH1oaA1zVsASeIQXTLhpJaAbQs/eyjCTWOWlj8m2\\nxOZ1EEwCVY3i7Ik32pkwprD/gY3goz1h1byiJnSss+wAi4pNiu+X62UEVM7bnQc=\\n=8E7n\\n-----END
  PGP MESSAGE-----\\n\",\n\t\t\t\t\"fp\": \"A883B02E14B7EAF0A6BDAF69ECD23C5916E31D4F\"\n\t\t\t}\n\t\t],\n\t\t\"encrypted_regex\":
  \"^(data|stringData)$\",\n\t\t\"version\": \"3.7.0\"\n\t}\n}"
```

Please note this is something I have basically no reason to expect to work. I am not sure what basis there is to expect `.sops.yaml` to be honored, as the decryption is implemented in go code, and reading over it, the only place I see `.sops.yaml` mentioned is in our SOPS guide, so unless the SOPS go module is doing something to read it in, somewhere else that I didn't see, I guess my instructions in `.sops.yaml` would be ignored at runtime.

Maybe it will work clientside or in CI, but not here. (Maybe this is an enhancement request?)

I'm still using Flux 0.10.0, haven't read the 0.11.0 release notes yet, so not sure if this would have been addressed by the latest release, either.

**Reply by stefanprodan** [reply]
The kustomize-controller can’t shell-out to random binaries and its file system is readonly due to security concerns so using secretsGenerator is not an option.

**Reply by kingdonb** [reply]
@stefanprodan do you think the instructions provided in `.sops.yaml` are ignored by our go sops client routines?

That's what it looks like from here. That kustomize controller will only decrypt Kubernetes yamls as it is applying them.

The closest mention in SOPS guide comes to clarifying this is this hint:

> Hint

> Note that you should encrypt only the data or stringData section. Encrypting the Kubernetes secret metadata, kind or apiVersion is not supported by kustomize-controller.

I don't know how we can make this more prominent, (it's already in a call-out). Sops is not a random binary, though. If we put `.sops.yaml` in the tutorial, people are going to have the expectation that they can put their own configurations in there too.

**Reply by stefanprodan** [reply]
> If we put .sops.yaml in the tutorial, people are going to have the expectation that they can put their own configurations in there too.

I don’t understand how is this related to decryption?

**Reply by kingdonb** [reply]
We should make it clearer that `.sops.yaml` is only for the end-user client, and doesn't perform any configuration for Kustomize controller. To me it looks like it's not supported, only data and stringData can be decrypted for now.

**Reply by stefanprodan** [reply]
The decryption config is not in `.sops.yaml` but in the metadata injected by SOPS in each file.

**Reply by apeschel** [reply]
@stefanprodan SOPS provides a Go library for decrypting SOPS files.

In theory, if one wanted to support SecretsGeneators with Flux, one could import the SOPS library and decrypt all the YAML files in a repository with the Kustomize controller as a preliminary step, before doing normal operations.

Your thoughts?

https://pkg.go.dev/go.mozilla.org/sops/decrypt?utm_source=godoc

**Comment by bernardhalas** [comment]
I'm coming here from the `git-crypt` and `fluxv1` world, where we used to encrypt a specific folder in our repo, containing all the sensitive information. We haven't found a way how to move forward with `fluxv2`. Just stating our use-case. The rest has been already mentioned.

**Comment by apeschel** [comment]
@stefanprodan SOPS provides a Go library for decrypting SOPS files.

In theory, if one wanted to support SecretsGeneators with Flux, one could import the SOPS library and decrypt all the YAML files in a repository with the Kustomize controller as a preliminary step, before doing normal operations.

Your thoughts?

https://pkg.go.dev/go.mozilla.org/sops/decrypt?utm_source=godoc

**Reply by stefanprodan** [reply]
@apeschel see https://github.com/fluxcd/kustomize-controller/pull/329

**Comment by stefanprodan** [comment, marked as answer]
Starting with flux2 **v0.13.2** we support decryption of Kubernetes secretes generated by Kustomize `secretGenerator`. 
Docs here: https://fluxcd.io/docs/components/kustomize/kustomization/#kustomize-secretgenerator

Thanks @bob-rohan for implementing this feature in kustomize-controller 🥇

**Reply by sobi3ch** [reply]
This is great addition! Is there any example how to use it with env files? I cannot make it to work.

**Reply by XtremeAI** [reply]
Thanks for update @stefanprodan! 
However... I am unsure what happens...
I've upgraded to 0.13.3 and did:
```
cat /tmp/secret-values.yaml | sops --config .sops.yaml -e /dev/stdin > secret-values.yaml.encrypted
```
```
configurations:
  - kustomize-config.yaml

secretGenerator:
 - name: secret-values
   files:
   - values.yaml=secret-values.yaml.encrypted
```
I need to use it as reference to values in Flux's HelmRelease. So I do
```
kind: HelmRelease
...
spec:
...
  valuesFrom:   
    - kind: Secret
      name: secret-values # will be customized based on Secret generated name by Kustomize
```
Looking at the secret:
```
k get secrets -n flux-system secret-values-59m92fc2hg -o yaml | yq e '.data."values.yaml"' - | base64 -d
{
        "data": "ENC[AES256_GCM .... 
```
Obviously the generated secret got applied to the cluster with the data still encrypted and unusable by HelmRelease... 
Guys, @bob-rohan, am I missing something?

**Reply by bob-rohan** [reply]
Looks reasonable. Very similar to my setup. 

Worth to check you've applied the sops decrypt to the Kustomization CRD? And appropriate means for decryption has been supplied to the kustomize-controller?

**Reply by XtremeAI** [reply]
Well, yeah...
The sops decrypt config is configured on the Kustomization object of Flux and the decryption of static secrets works well:
```
apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
kind: Kustomization
...
spec:
...
  decryption:
    provider: sops
``` 
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
...
resources:
- ./secrets.yaml
```
the `secrets.yaml`:
```
apiVersion: v1
data:
    values.yaml: ENC[AES256_GCM,data:...
kind: Secret
```
gets decrypted and applied to the cluster:
```
apiVersion: v1
data:
  values.yaml: ZW5...
  ```

**Reply by stefanprodan** [reply]
You need to configure kustomize to patch the secret suffix, example here https://fluxcd.io/docs/guides/helmreleases/#refer-to-values-in-configmaps-generated-with-kustomize

**Reply by XtremeAI** [reply]
Something like this ? 
```
nameReference:
- kind: Secret
  version: v1
  fieldSpecs:
  - path: spec/valuesFrom/name
    kind: HelmRelease

```
I did and I do get the secret applied to the cluster however the content of the secret is still encrypted .

**Comment by stealthybox** [comment]
> It would be super cool if Kustomize controller would sops decrypt the values before it does kustomize build so we could use generated secrets!

This is the main issue here.

The kustoimize-controller does decryption by first making the filesystem, doing the `kustomize build`, and then iterating over the objects with a `sops decrypt`.

This makes it impossible to build Secrets from multiple encrypted patches or via `secretGenerator` -- both of which are powerful delegation mechanisms.

We should change this to decrypt the filesystem before the `kustomize build`.

Relevant code:
https://github.com/fluxcd/kustomize-controller/blob/871c2a1/controllers/kustomization_controller.go#L522-L534

**Reply by stefanprodan** [reply]
We do support secretGenerator, see the docs here: https://fluxcd.io/docs/components/kustomize/kustomization/#kustomize-secretgenerator

**Reply by stefanprodan** [reply]
> We should change this to decrypt the filesystem before the kustomize build.

Many keys can be used in the same repo, e.g. one key per cluster, or one key per tenant, etc. How would the controller know which files to try to decrypt. Also the location of the files included in an overlay can be anywhere, even remote.

**Reply by XtremeAI** [reply]
@stefanprodan, @bob-rohan : 

Again, I am on Flux 0.13.3 and tried to follow the docs you provided. 

> SOPS encrypted data can be stored as a base64 encoded Secret, which enables the use of Kustomize secretGenerator ...

Do you mean on Source (Git) or on the cluster? 

> The kustomize-controller scans the values of Kubernetes Secrets, and when it detects that the values are SOPS encrypted, it decrypts them before applying them on the cluster.

The result that is applied to the cluster in my case is a secret with encrypted data in it:
```
k get secret token-mtb64c7725 -o yaml | yq e '.data.token' - | base64 -d
{
        "data": "ENC[AES256_GCM,data
        ....
```
is that what you intend to support or should I expect it to be decrypted? I am confused. 

I don't observe that the values are decrypted... could you please clarify?

**Reply by XtremeAI** [reply]
> > It would be super cool if Kustomize controller would sops decrypt the values before it does kustomize build so we could use generated secrets!
> 
> We should change this to decrypt the filesystem before the `kustomize build`.
> 
> Relevant code:
> https://github.com/fluxcd/kustomize-controller/blob/871c2a1/controllers/kustomization_controller.go#L522-L534

Yep. Makes sense...

**Comment by stefanprodan** [comment]
With [SOPS: Decrypt dotenv files used in kustomize secret generator](https://github.com/fluxcd/kustomize-controller/pull/463) we now cover all kustomize `secretGenerator` options in Flux.

**Reply by BlackBsd** [reply]
I am still having the same issue as mentioned above.  The generated secret from kustomize ends up looking like a sops encrypted secret once base64 decoded.  

```
kubectl -n default get secret my-docker-config-secret -o yaml | yq e '.data.".dockerconfigjson"' - | base64 -d
{
        "auths":  {
            "https://index.docker.io/v1/": {
                "auth": "ENC[AES256_GCM,data.....]",
                "email": "ENC[AES256_GCM,data.....]"
        ....
        "sops": {
        ....
        }
}
```


I am trying to follow the documentation at https://fluxcd.io/flux/components/kustomize/kustomization/#kustomize-secretgenerator.

I would expect that the final secret created would look decrypted and only hold the proper values?

**Reply by BlackBsd** [reply]
Just one more follow up after some testing.  I think its working properly now.

TLDR:  Please verify that you have set the decryption provider for all `Kustomization`s your creating that have secrets in them.

In my case, I only had the decryption provider set up on the `Kustomization` file under the `flux-system` folder, i.e `./k8s/clusters/dev/flux-system`.  and not in my main `Kustomization` for my infrastructure.


Here is copy, which is located in `./k8s/clusters/dev/infrastructure`
```
apiVersion: kustomize.toolkit.fluxcd.io/v1beta2
kind: Kustomization
metadata:
  name: infrastructure
  namespace: flux-system
spec:
  sourceRef:
    kind: GitRepository
    name: flux-system
  decryption:
    provider: sops
    secretRef:
      name: sops-gpg
  path: ./k8s/infra/dev
```

I also needed to have the decryption provider set up for every sourceRef that holds secrets.
